toolchain/check/context.cpp

```diff
@@ -279,6 +279,228 @@ auto Context::ParamOrArgEnd(Parse::NodeKind start_kind) -> SemIR::NodeBlockId {
+        case SemIR::BuiltinKind::BoolType:
+        case SemIR::BuiltinKind::IntegerType:
+        case SemIR::BuiltinKind::FloatingPointType:
+        case SemIR::BuiltinKind::NamespaceType:
+        case SemIR::BuiltinKind::FunctionType:
+          return set_copy_representation(type_id);
+
+        case SemIR::BuiltinKind::StringType:
+          // TODO: Decide on string value semantics. This should probably be a
+          // custom value representation carrying a pointer and size or
+          // similar.
+          return set_pointer_representation(type_id);
+      }
+      llvm_unreachable("All builtin kinds were handled above");
+    }
+
+    case SemIR::ArrayType::Kind:
+      // For arrays, it's convenient to always use a pointer representation,
+      // even when the array has zero or one element, in order to support
+      // indexing.
+      return set_pointer_representation(type_id);
+
+    case SemIR::StructType::Kind: {
+      auto fields =
+          semantics_ir().GetNodeBlock(node.As<SemIR::StructType>().fields_id);
+      if (fields.empty()) {
+        return set_empty_representation();
+      }
+
+      // Find the value representation for each field, and construct a struct
+      // of value representations.
+      llvm::SmallVector<SemIR::NodeId> value_rep_fields;
+      value_rep_fields.reserve(fields.size());
+      bool same_as_object_rep = true;
+      for (auto field_id : fields) {
+        auto field = semantics_ir().GetNodeAs<SemIR::StructTypeField>(field_id);
+
+        // A struct is complete if and only if all its fields are complete.
+        auto field_value_rep =
+            semantics_ir().GetValueRepresentation(field.type_id);
+        if (field_value_rep.kind == SemIR::ValueRepresentation::Unknown) {
+          // TODO: If the field type might have become complete after we formed
+          // it, we should attempt to complete its type.
+          return false;
+        }
+        if (field_value_rep.type_id != field.type_id) {
+          same_as_object_rep = false;
+          field.type_id = field_value_rep.type_id;
+          field_id = AddNode(field);
+        }
+        value_rep_fields.push_back(field_id);
+      }
+
+      auto value_rep = same_as_object_rep
+                           ? type_id
+                           : CanonicalizeStructType(
+                                 node.parse_node(),
+                                 semantics_ir().AddNodeBlock(value_rep_fields));
+      if (fields.size() == 1) {
+        // The value representation for a struct with a single field is a struct
+        // containing the value representation of the field.
+        // TODO: Consider doing the same for structs with multiple small fields.
+        return set_copy_representation(value_rep);
+      }
+      // For a struct with multiple fields, we use a pointer representation.
+      return set_pointer_representation(value_rep);
+    }
+
+    case SemIR::TupleType::Kind: {
```

Comment:
Seems like it'd be worth at least commoning this with structs. And some parts might make sense in common even with arrays.

But all of that seems like a future-patch, so maybe just a TODO?

```suggestion
    case SemIR::TupleType::Kind: {
      // TODO: Extract and share code with structs and maybe arrays.
```